src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,284 @@
+public class GlobalGlaciationEvent : IWorldEffect
+{
+    private const string TemplateBiomeForIceChunks = "ice_shelf";
+    private const string Prefix = "globalGlaciation_";
+    private const string Background = "iceshelf";
+
+    private static readonly string[] IceChunksConfigurations =
+        ["iceShard", "iceChunkSmall", "iceChunkBig", "iceSnowflake"];
+
+    [JsonProperty]
+    private readonly XoShiRo256starstar random;
+
+    [JsonProperty]
+    private readonly List<int> modifiedPatchesIds = new();
+
+    [JsonProperty]
+    private bool hasEventAlreadyHappened;
+
+    /// <summary>
+    ///   Tells how many generations the event will last. "-1" means that it hasn't started at all.
+    ///   "0" means it has finished, and it won't happen again
+    /// </summary>
+    [JsonProperty]
+    private int generationsLeft = -1;
+
+    [JsonProperty]
+    private int generationsToTrigger = -1;
+
+    [JsonProperty]
+    private int eventDuration;
+
+    [JsonProperty]
+    private GameWorld targetWorld;
+
+    public GlobalGlaciationEvent(GameWorld targetWorld, long randomSeed)
+    {
+        this.targetWorld = targetWorld;
+        random = new XoShiRo256starstar(randomSeed);
+    }
+
+    [JsonConstructor]
+    public GlobalGlaciationEvent(GameWorld targetWorld, XoShiRo256starstar random)
+    {
+        this.targetWorld = targetWorld;
+        this.random = random;
+    }
+
+    public void OnRegisterToWorld()
+    {
+    }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        if (hasEventAlreadyHappened)
+            return;
+
+        if (generationsLeft > 0)
+            generationsLeft -= 1;
+
+        // Mark patches with the event icon while the event lasts
+        if (generationsLeft > 0)
+            MarkPatches(totalTimePassed);
+
+        if (generationsLeft == -1)
+        {
+            TryToTriggerEvent(totalTimePassed);
+        }
+        else if (generationsLeft == 0)
+        {
+            FinishEvent();
+        }
+    }
+
+    private bool IsSurfacePatch(Patch patch)
+    {
+        return patch.Depth[0] == 0 && patch.BiomeType != BiomeType.Cave;
+    }
+
+    private void MarkPatches(double totalTimePassed)
+    {
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (IsSurfacePatch(patch))
+            {
+                patch.AddPatchEventRecord(WorldEffectVisuals.GlobalGlaciation, totalTimePassed);
+            }
+        }
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (generationsToTrigger == -1)
+        {
+            if (!AreConditionsMet())
+                return;
+
+            generationsToTrigger = Constants.GLOBAL_GLACIATION_HEADS_UP_DURATION;
+            return;
+        }
+
+        switch (generationsToTrigger)
+        {
+            case > 0:
+                LogHeadUpEventWarning();
+                generationsToTrigger -= 1;
+                break;
+            case 0:
```

Comment:
Seems a bit strange to me to use a switch statement for 2 possible branches, an if would be a lot shorter.